# no-duplicate-keys

Disallow duplicate keys in JSON objects.

## Background

This rule warns when there are two keys in an object with the same text.

## Rule Details

Examples of **incorrect** code for this rule:

```jsonc
/* eslint json/no-duplicate-keys: "error" */

```

Examples of **correct** code for this rule:

```jsonc
/* eslint json/no-duplicate-keys: "error" */

```

## Options

## When Not to Use It
