# no-duplicate-keys

Warns when there are two keys in an object with the same text.

## Background

## Rule Details

Examples of **incorrect** code for this rule:

```jsonc
/* eslint json/no-duplicate-keys: "error" */

```

Exmples of **correct** code for this rule:

```jsonc
/* eslint json/no-duplicate-keys: "error" */

```

## Options

## When Not to Use It
