# no-unnormalized-keys

## Background

## Rule Details

Examples of **incorrect** code for this rule:

```jsonc
/* eslint json/no-unnormalized-keys: "error" */

```

Exmples of **correct** code for this rule:

```jsonc
/* eslint json/no-unnormalized-keys: "error" */

```

## Options

## When Not to Use It
