# sort-keys

Require JSON object keys to be sorted.

## Background

This rule warns when keys are not in the specified order. Based on the ESLint [`sort-keys`](https://eslint.org/docs/latest/rules/sort-keys) rule.

## Rule Details

Examples of **incorrect** code for this rule:

```jsonc
/* eslint json/sort-keys: "error" */

```

Examples of **correct** code for this rule:

```jsonc
/* eslint json/sort-keys: "error" */

```

## Options

## When Not to Use It
