# top-level-interop

## Background

## Rule Details

Examples of **incorrect** code for this rule:

```jsonc
/* eslint json/top-level-interop: "error" */

```

Exmples of **correct** code for this rule:

```jsonc
/* eslint json/top-level-interop: "error" */

```

## Options

## When Not to Use It
